docs: update db migration, environment and restful api practicesC (#56)
dwarvesf/playbook: engineering/backend/restful.md

## Convention

> - We prefer Swagger for REST API documentation. Alternative can be postman

Reviewer: ```suggestion
- We prefer Swagger for REST API documentation.
```
we can use a tool to generate postman file from swagger

## Filter, Search and Sort
Use query parameters for advanced filtering, sorting & searching.

###Pagination

**Offset pagination:**

- `page` - current page wanted to fetch
- `size` - number of items per page

- The API should return `totalItems` and `totalPages`

Example:
```
GET /companies?page=23&size=50
```

**Cursor pagination**

- `cursor` - is the reference element in the list
- `size` - number of items per page

> - The API should return `previousCursor` and `nextCursor`

Reviewer: should describe how we structure the cursor from the backend and use from client

> - `cursor` - is the reference element in the list. Best brattice for choosing the cursor is only sort by unique sequential column, create new column which is concatenate columns if you have to order by multiple none unique columns.

Reviewer: ```suggestion
- `cursor` - is the reference element in the list. The best practice for choosing the cursor is only sorting by a unique sequential column, and creating a new concatenate column if you have to order by multiple none unique columns.
```

```
GET /companies?cursor=12&size=50&cursorDirection=
```

### Sort:

- Validate all sort able fields
- By default sort in ascending order
- Use  `minus(-)` prefix to indicate sort in descending order

Example
```
# sort by created_at in descending order
GET /companies?sort=-created_at

# sort by created_at in ascending order
GET /companies?sort=created_at
```

### Filter:

- Validate all enums values
- Use repeated query keys for multiple values
- Use RFC339 time format or unix timestamp for date time filter 

```

> # filter by category
> GET /companies?category=software
> 
> # filter by locations
> GET /companies?location=saigon&location=hanoi
> 
> # Filter by time range

Reviewer: It makes me confused about the markdown convention. As we knew, One # character for Heading 1. We should update the way we index the headings.
```suggestion
#### Filter by category
GET /companies?category=software

#### Filter by locations
GET /companies?location=saigon&location=hanoi

#### Filter by time range
```

GET /companies?start=1662873449&end=1662873449

> GET /companies?start=2022-09-10T05:10:55.308Z&end=2022-09-11T05:10:55.308Z

Reviewer: how can we filter 2+ DateTime fields with this convention?
